### ISSUE TITLE
dask data partitions may break the groups with qid for ranker

### ISSUE THREAD

**Issue by weitian** [dmlc/xgboost#8691]
for ranker, training data are organized in groups.
Currently xgboost-dask uses qid, but it does not handle the group boundary correctly while partitions the data.

for example, a group has 20 rows, but 8 rows could go to end of partition 1 and 12 rows go to beginning of partition 2.

**Comment by trivialfis** [comment]
As long as the order within the same worker is kept, this should be fine. All partitions within the same worker are joined.

We can calculate the lambda gradient on each worker even if the query group is scattered among many different workers. The accuracy may change depending on the scattering, but I don't see a practical way to ensure all data within the same query group is allocated to a single worker. For instance, what if there's only one group?

**Comment by weitian** [comment]
Only one group is a rare corner case for ranker.

Use our data as an example, we have 10 million plus groups in training data which is the reason we want to use Dask to partition them.

Currently the workaround I use for my self is that, I blindly drop the first and last groups in a partitions no matter they cross partitions or not. Since dropping a few groups out of millions does not affect the training at all.

**Comment by trivialfis** [comment]
I don't understand why do you need to do that. Please elaborate.

As mentioned, if the partitions are on the same worker, then they are joined into one. If the partitions are scattered among different workers, then the gradient can still be calculated independently, we will sum the gradient among workers during tree split.

**Comment by weitian** [comment]
@trivialfis , for [rank:pairwise] ranking candidates are in groups. When Dask split data for workers, a group may be split into two partitions for two workers. For example, if there is a group having 15 candidates, 10 candidates in partition 1 which goes to worker 1, and another 5 candidates in partition 2 which goes to worker 2.

My question is that, can 'sum the gradient histogram among workers' handle the above situation ?

**Comment by trivialfis** [comment]
We calculate the gradient for each sample by comparing samples in each group. If the group is split among different workers, the gradient for samples in the group will be different due to lesser document pairs that can be used for comparison on each worker. However, the changed gradient is still theoretically correct even if it's not optimal, it's just having lesser samples in a query group.

The gradient histogram doesn't need to take the group into consideration, it's an aggregation of gradients from each sample.
